tracing-syslog/src/syslog.rs

```diff
@@ -0,0 +1,246 @@
+    // format string and formatting placeholders e.g. %s must correspond to
+    // one of the variable-length arguments. By construction, the format string
+    // is nul-terminated, and the only string formatting placeholder corresponds
+    // to `msg.as_ptr()`, which is a valid, nul-terminated string in C world
+    // because `msg` is a `CStr`.
+    unsafe { libc::syslog(priority.0, "%s\0".as_ptr().cast(), msg.as_ptr()) }
+}
+
+/// [`Subscriber`](tracing_subscriber::Subscribe) that logs to `syslog`.
+///
+/// # Examples
+/// Initializing a global [`Collector`](tracing_core::Collect) that logs to `syslog` with
+/// an identity of `example-program` and the default `syslog` options and facility:
+/// ```
+/// use tracing_syslog::Syslog;
+/// use tracing_subscriber::{Registry, subscribe::CollectExt};
+/// let identity = std::ffi::CStr::from_bytes_with_nul(b"example-program\0").unwrap();
+/// let (options, facility) = Default::default();
+/// let collector = Registry::default().with(Syslog::new(identity, options, facility));
+/// tracing::collect::set_global_default(collector).unwrap();
+/// ```
+pub struct Syslog {
+    /// Identity e.g. program name. Referenced by syslog, so we store it here to
+    /// ensure it lives until we are done logging.
+    #[allow(dead_code)]
+    identity: Cow<'static, CStr>,
+    facility: Facility,
+}
+
+impl Syslog {
+    /// Creates a [`Subscriber`](tracing_subscriber::Subscribe) that logs to `syslog`.
+    ///
+    /// # Examples
+    /// Creating a `syslog` subscriber with an identity of `example-program` and
+    /// the default `syslog` options and facility:
+    /// ```
+    /// use tracing_syslog::Syslog;
+    /// let identity = std::ffi::CStr::from_bytes_with_nul(b"example-program\0").unwrap();
+    /// let (options, facility) = Default::default();
+    /// let subscriber = Syslog::new(identity, options, facility);
+    /// ```
+    pub fn new(
+        identity: impl Into<Cow<'static, CStr>>,
+        options: Options,
+        facility: Facility,
+    ) -> Self {
+        let identity = identity.into();
+        // SAFETY: identity will remain alive until the returned struct's fields
+        // are dropped, by which point `closelog` will have been called by the
+        // `Drop` implementation.
+        unsafe { libc::openlog(identity.as_ptr(), options.0, facility as libc::c_int) };
+        Syslog { identity, facility }
+    }
+}
+
+impl Drop for Syslog {
+    fn drop(&mut self) {
+        unsafe { libc::closelog() };
+    }
+}
+
+impl<C> Subscribe<C> for Syslog
+where
+    C: Collect + for<'span> LookupSpan<'span>,
+{
+    fn on_event(&self, event: &Event, _ctx: Context<C>) {
+        use std::cell::RefCell;
+        thread_local! { static BUF: RefCell<Vec<u8>> = RefCell::new(Vec::with_capacity(256)) }
+
+        BUF.with(|buf| {
+            let mut buf = buf.borrow_mut();
+
+            // Record event fields
+            event.record(&mut EventVisitor(&mut buf));
```

Comment:
Right now, this will only record the event's fields --- the formatted message won't include the current span context, or any metadata from the event (such as the file and line number, module path, log target, etc) or additional data like thread names/IDs. That's kind of a bummer. It would be nice if we could allow users to customize how the event is formatted.

We could reproduce all the logic `tracing-subscriber`'s `fmt` module has for formatting events, but I think that would kind of be a shame. Instead, I think it would be great if that code could be reused, so that users can use the same interface for configuring event formats for `tracing-syslog` and `fmt`.

I think there are a couple ways to achieve this, but both of them will require some changes in `tracing-subscriber` (which I've planned on making anyway). Here's what I'm thinking:

1. Rather than having the `tracing-syslog` crate provide a `Subscribe` implementation, it could provide a type implementing `MakeWriter` instead. That way, a syslog writer could be passed into `tracing-subscriber`'s `fmt` layer, and the syslog crate would _just_ be responsible for writing data to the syslog, and all event formatting could be handled by `fmt`. That would let users write code like this:
   ```rust
   let (options, facility) = Default::default();
   let syslog = Syslog::new(identity, options, facility);
   tracing_subscriber::fmt()
       .with_writer(syslog)
       .event_format(/* pass in some arbitrarily configured user event formatter here */)
       .init();
   ```

   The `Write` implementation would probably look something like this:

   ```rust
   pub struct SyslogWriter {
       flushed: bool,
       facility: Facility,
   };
   
   thread_local! { static BUF: RefCell<Vec<u8>> = RefCell::new(Vec::with_capacity(256)) }
   
   impl io::Write for SyslogWriter {
       fn write(&mut self, bytes: &[u8]) -> io::Result<usize> {
           BUF.with(|buf| {
               buf.borrow_mut().extend(bytes)
           });
           Ok(())
       }
   
       fn flush(&mut self) -> io::Result<()> {
           BUF.with(|buf {
               let mut buf = buf.borrow_mut();
   
               // Append nul-terminator
               buf.push(0);
   
               // Log message
               let priority = Priority::new(self.facility, /* where to get the level from??? */);
               let msg =
                   CStr::from_bytes_with_nul(&buf).expect("logs free of interior nul-terminators");
               syslog(priority, &msg);
   
               // Clear buffer
               buf.clear();
           });
           self.flushed = true;
           Ok(())
       }
   }
   
   impl Drop for SyslogWriter {
       fn drop(&mut self) {
           if !self.flushed {
               let _ = self.flush();
           }
       }
   }
   ```
  As you've probably noticed, there's one big issue with this approach. The `MakeWriter` trait doesn't currently know anything about the event (or span) that's being written, so there's no way to determine the appropriate severity for the call to `syslog(3)`. 

   However, I've been considering adding a new method to `MakeWriter`, `writer_for_metadata` (or similar, name TBD) that takes a `Metadata` argument and allows the `MakeWriter` implementor to configure a writer based on the metadata of the span/event being written. There would be a default impl that just calls the `MakeWriter::make_writer` method, but writers that have specialized behavior based on metadata could override it to opt in. This would allow several things that people might want to do: automatically use different ANSI color codes based on level, write logs from different targets to different files, write logs from the WARN and ERROR levels to stderr instead of stdout, and so on. In addition, it would allow the syslog writer to be configured with the event's level, so that it can determine the correct priority.

  If you think this seems like a good approach, I can put together a PR for adding metadata awareness to `MakeWriter`, and you can rebase once that merges?
2. Alternatively, we could continue implementing `Subscribe` for the syslog crate, but reuse `tracing-subscriber::fmt`'s `FormatEvent` trait. Right now, this is difficult, because `FormatEvent` has an argument that can't be publicly constructed; see https://github.com/tokio-rs/tracing/issues/1039 etc. I'd like to rework that interface so it can be reused in other crates, but this might be a bit more involved than the first approach.

Personally, I think that option 1, exposing syslog as a specialized `MakeWriter` that a `fmt` subscriber can be configured to write to, seems like it might be the best approach. The syslog crate could continue exposing an API for just getting a subscriber, that returns the `fmt` subscriber configured to write to syslog, so that there's a one-function-call API for writing to syslog. What do you think?

Reply:
I think option 1 makes sense given it's possible to disable the emission of things that `syslog` adds on itself (timestamp and level are the only two I think, and these look like they can both be disabled in `fmt`).

Reply:
Do you have an estimate as to when the PR might be filed and/or land?

Reply:
I was hoping to get to it this afternoon but some other things came up — I'll probably have it tomorrow.

Reply:
@max-heller I've opened PR  #1141 to add `MakeWriter::make_writer_for` with metadata.

Reply:
Got around to switching over to the `MakeWriter` approach!